runtime/src/main/scala/akka/grpc/internal/ServerReflectionImpl.scala

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright (C) 2019-2020 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.grpc.internal
+
+import com.google.protobuf.Descriptors.FileDescriptor
+
+import akka.NotUsed
+import akka.stream.scaladsl._
+
+import _root_.grpc.reflection.v1alpha.reflection._
+
+class ServerReflectionImpl private (fileDescriptors: Map[String, FileDescriptor], services: List[String])
+    extends ServerReflection {
+  import ServerReflectionImpl._
+
+  def serverReflectionInfo(in: Source[ServerReflectionRequest, NotUsed]): Source[ServerReflectionResponse, NotUsed] = {
+    in.map(req => {
+      import ServerReflectionRequest.{ MessageRequest => In }
+      import ServerReflectionResponse.{ MessageResponse => Out }
+
+      val response = req.messageRequest match {
+        case In.Empty =>
+          Out.Empty
+        case In.FileByFilename(fileName) =>
+          val list = fileDescriptors.get(fileName).map(_.toProto.toByteString).toList
+          Out.FileDescriptorResponse(FileDescriptorResponse(list))
+        case In.FileContainingSymbol(symbol) =>
+          val list = findFileDescForSymbol(symbol, fileDescriptors).map(_.toProto.toByteString).toList
+          Out.FileDescriptorResponse(FileDescriptorResponse(list))
+        case In.FileContainingExtension(ExtensionRequest(container, number)) =>
+          val list = findFileDescForExtension(container, number, fileDescriptors).map(_.toProto.toByteString).toList
+          Out.FileDescriptorResponse(FileDescriptorResponse(list))
+        case In.AllExtensionNumbersOfType(container) =>
+          val list = findExtensionNumbersForContainingType(container, fileDescriptors) // TODO should we throw a NOT_FOUND if we don't know the container type at all?
+          Out.AllExtensionNumbersResponse(ExtensionNumberResponse(container, list))
+        case In.ListServices(_) =>
+          val list = services.map(s => ServiceResponse(s))
+          Out.ListServicesResponse(ListServiceResponse(list))
+      }
+      // TODO Validate assumptions here
+      ServerReflectionResponse(req.host, Some(req), response)
+    })
+  }
+}
+object ServerReflectionImpl {
+  import scala.collection.JavaConverters._
+
+  def apply(fileDescriptors: Seq[FileDescriptor], services: List[String]): ServerReflectionImpl =
+    new ServerReflectionImpl(
+      (AdditionalDescriptors ++ fileDescriptors).map(fd => fd.getName -> fd).toMap,
+      ServerReflection.name +: services)
+
+  val AdditionalDescriptors =
+    flattenDependencies(ReflectionProto.javaDescriptor).distinct
+
+  private def flattenDependencies(descriptor: FileDescriptor): List[FileDescriptor] = {
+    descriptor :: descriptor.getDependencies.asScala.toList.flatMap(flattenDependencies)
+  }
+
+  def splitNext(name: String): (String, String) = {
```

Comment:
internal API most of these methods (and fields) or the object itself (apply is called from user code?)

Reply:
user code shouldn't need to refer to the `Impl` at all, the `ServerReflection` in javadsl/scaladsl should be enough

Reply:
ok, so private[akka] the and add `@InternalApi` annotation to the object then rather